README.md
# ELS - Enterprise Linux/EuroLinux Sources - libotf
 
## Usage:
  Checkout branch or tag. Use `make sources` to recreate  SOURCE0 archive.


src/ChangeLog
2012-09-18  Kenichi Handa  <handa@gnu.org>

	* Version 0.9.13 released.

2012-04-27  Kenichi Handa  <handa@m17n.org>

	* configure.ac: Change versions to 0.9.13.

2012-04-23  Kenichi Handa  <handa@m17n.org>

	* src/otfdrive.c (get_class_def_auto): Return 0 if C is negative.

2011-02-11  Kenichi Handa  <handa@m17n.org>

	* src/otfopen.c (read_anchor_array): Do not read an anchor if its
	offset is zero.
	(read_ligature_attach): Fix return code on error.
	(read_ligature_array): Do not read an anchor if its offset is
	zero.
	(read_lookup_subtable_gpos): Fix typo.

	* example/otfdump.c (indent_spaces): Delete it.
	(IPRINT): Do not use indent_spaces.
	(dump_anchor): Do nothing if anchor->offset is zero.

2010-11-05  Elias Pipping  <pipping.elias@googlemail.com>  (tiny change)

	* example/Makefile.am (bin_PROGRAMS): Add otfview conditionally.

	* configure.ac: Add AM_CONDITIONAL for X11_UNAVAILABLE.

2010-10-04  Kenichi Handa  <handa@m17n.org>

	* Version 0.9.12 released.

2010-09-22  Kenichi Handa  <handa@m17n.org>

	* src/otfopen.c (OTF_check_features): If language is specified but
	not found, return 0 instead of using the default langsys.

2010-09-16  Kenichi Handa  <handa@m17n.org>

	* src/internal.h: Include alloca.h if available.

	* example/otftobdf.c: Include alloca.h if available.

	* example/otfview.c: Include alloca.h if available.

	* example/otflist.c: Include alloca.h if available.

2010-09-13  Martin Gieseking  <martin.gieseking@uos.de>  (tiny change)

	* example/otftobdf.c (main): Fix checking number of arguments.

2010-09-10  Kenichi Handa  <handa@m17n.org>

	* src/otfopen.c (OTF_check_features): If LANGUAGE is specified but
	not found in a font, return 0.

2010-05-18  Kenichi Handa  <handa@m17n.org>

	* example/otfview.c (create_otf_script_widgets): Create a box
	widget for each script in a limited-width box.

2010-04-21  Kenichi Handa  <handa@m17n.org>

	* configure.ac: Versions changed to 0.9.12.

	* src/otfopen.c (OTF_check_features): Even if no feature is
	specified, check script support.

	* src/otf.h (LIBOTF_VERSION): Changed to 0.9.12
	(LIBOTF_RELEASE_NUMBER): Changed to 12.

2010-03-30  Kenichi Handa  <handa@m17n.org>

	* configure.ac: Versions changed to 0.9.11.

2010-03-23  Kenichi Handa  <handa@m17n.org>

	* src/Makefile.am (libotf_la_SOURCES): Include internal.h.

2010-03-19  Kenichi Handa  <handa@m17n.org>

	* src/Makefile.am (libotf_la_SOURCES): Exclude otferror.h.

	* configure.ac: Versions changed to 0.9.11pre.

2010-01-22  Kenichi Handa  <handa@m17n.org>

	* src/otfdrive.c (lookup_gsub): Change the meaning of the arg
	alternate_subst.

2009-12-17  Kenichi Handa  <handa@m17n.org>

	* src/otfdrive.c (gstring_subst): Update g->MarkAttachGlass.
	(lookup_gsub): Update g->GlyphClass and g->MarkAttachGlass.
	(OTF_drive_gdef): Don't return error even if GDEF table doesn't
	exist.

2009-11-30  Kenichi Handa  <handa@m17n.org>

	* src/otfdrive.c (OTF_drive_gsub_internal): Fix previous change.
	(OTF_drive_gpos_internal): Likewise.

2009-11-27  Kenichi Handa  <handa@m17n.org>

	* src/otfdrive.c (gstring_subst): Set positioning_type to 1.
	(lookup_gsub): If a feature is applied to a glyph, set its
	positioning_type to 1.
	(OTF_drive_gsub_internal): Only if a feature is applied to a
	glyph, set its feature log.
	(OTF_drive_gpos_internal): Likewise.

2009-11-24  Kenichi Handa  <handa@m17n.org>

	* src/otf.h (OTF_drive_gsub_with_log, OTF_drive_gpos_with_log):
	Extern them.
	(OTF_get_drive_log): Delete extern.

	* src/internal.h (OTF_DriveLog): Delete it.
	(set_drive_log, get_drive_log): Delete externs.

	* src/otfopen.c (OTF_InternalData): Delete the member drive_log.
	(set_drive_log, get_drive_log): Delete them.

	* src/otfdrive.c (USHORT): New macro for "unsigned short".
	(prepare_drive_log): Delete it.
	(setup_lookup_flags): Change the type of lookup_flags.
	(lookup_gsub): Change the return value.
	(gstring_insert_for_gpos): Adjusted for the change of
	OTF_Glyph.positioning_type.
	(print_glyph_positioning): Likewise.
	(lookup_gpos): Likewise. Change the return value.
	(OTF_drive_gsub_internal): New arg "with_log".  Adjusted for the
	change of OTF_Glyph.positioning_type.
	(OTF_drive_gsub): Fix args to OTF_drive_gsub_internal.
	(OTF_drive_gsub_with_log): New function.
	(OTF_drive_gpos_internal): New arg "with_log".  Adjusted for the
	change of OTF_Glyph.positioning_type.
	(OTF_drive_gpos): Fix args to OTF_drive_gpos_internal.
	(OTF_drive_gpos2): Likewise.
	(OTF_drive_gpos_with_log): New function.
	(OTF_drive_gsub_alternate): Fix args to OTF_drive_gsub_internal.
	(OTF_iterate_gsub_feature): Adjusted for the change of
	OTF_Glyph.positioning_type.
	(OTF_get_drive_log): Delete it.

	* example/otfview.c (update_seq_area): Call
	OTF_driver_gsub_with_log and OTF_driver_gpos_with_log.  Show which
	features are applied.
	(get_features): Fix previous change.

2009-11-24  mule  <handa@m17n.org>

	* src/otferror.h: Removed.  All codes move to internal.h.

	* src/internal.h: New file.

	* src/otfdrive.c: Include internal.h instead of otferror.h.
	(prepare_drive_log): New function.
	(OTF_drive_gpos_internal): Make it static.
	(OTF_get_drive_log): New function.

	* src/otfopen.c: Include internal.h instead of otferror.h.
	(OTF_InternalData): New member drive_log.
	(set_drive_log, get_drive_log): New functions.

	* src/otf.h (OTF_get_drive_log): Extern it.

	* example/otfview.c (get_features): Adjusted for the change of
	FeatureRec->features.
	(FeatureProc): Don't re-order FeatureRec->features.

2009-11-18  Kenichi Handa  <handa@m17n.org>

	* example/otfview.c (update_seq_area): Fix previous change.

	* src/otfdrive.c (match_ids): New arg "direction".  Callers changed.
	(match_ids_backward): Delete it.
	(match_chain_ids): Adjusted for the change fo match_ids.
	(match_classes): New arg "direction".  Callers changed.
	(match_chain_classes): Adjusted for the change fo match_classes.
	(match_coverages): New arg "direction".  Callers changed.

src/example/otftobdf.c
/* otftobdf.c -- Generate BDF font from OpenType font.

Copyright (C) 2003, 2004
  National Institute of Advanced Industrial Science and Technology (AIST)
  Registration Number H15PRO167

This file is part of libotf.

Libotf is free software; you can redistribute it and/or modify it
under the terms of the GNU Lesser General Public License as published
by the Free Software Foundation; either version 2.1 of the License, or
(at your option) any later version.

Libotf is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library, in a file named COPYING; if not,
write to the Free Software Foundation, Inc., 59 Temple Place, Suite
330, Boston, MA 02111-1307, USA.  */

#include <stdio.h>

#include <ft2build.h>
#include FT_FREETYPE_H

#define DEFAULT_PIXEL_SIZE 16

FT_Face face;

/* Format MSG by FMT and print the result to the stderr, and exit.  */

#define FATAL_ERROR(fmt, arg)	\
  do {				\
    fprintf (stderr, fmt, arg);	\
    exit (1);			\
  } while (0)

char *registry;

void
dump_header (FT_Face face, char *foundry, int nchars, int pixel_size)
{
  int width = ((face->bbox.xMax - face->bbox.xMin)
	       * pixel_size / face->units_per_EM);
  int height = ((face->bbox.yMax - face->bbox.yMin)
		* pixel_size / face->units_per_EM);
  int x = face->bbox.xMin * pixel_size / face->units_per_EM;
  int y = face->bbox.yMin * pixel_size / face->units_per_EM;

  printf ("STARTFONT 2.1\n");
  printf ("FONT -%s-%s-%s-R-Normal--%d-%d-72-72-C-%d-%s\n",
	  foundry, face->family_name, face->style_name,
	  pixel_size, pixel_size * 10, pixel_size * 10, registry);
  printf ("SIZE %d 72 72\n", pixel_size);
  printf ("FONTBOUNDINGBOX %d %d %d %d\n", width, height, x, y);
  printf ("STARTPROPERTIES 2\n");
  printf ("FONT_ASCENT %d\n", y + height);
  printf ("FONT_DESCENT %d\n", -y);
  printf ("ENDPROPERTIES 0\n");
  printf ("CHARS %d\n", nchars);
}

void
dump_tailer ()
{
  printf ("ENDFONT\n");
}

void
dump_image (int pixel_size, int index, int code, int full)
{
  int err = FT_Load_Glyph (face, index, FT_LOAD_RENDER | FT_LOAD_MONOCHROME);
  int i,j;
  unsigned char *buf;
  FT_GlyphSlot glyph;
  int dwidth, x, y;
  
  if (err)
    return;
  glyph = face->glyph;
  if (glyph->bitmap.rows * glyph->bitmap.width == 0)
    return;
  printf ("STARTCHAR U+%04X\n", code);
  printf ("ENCODING %d\n", code);
  printf ("SWIDTH %d 0\n",
	  (int) (glyph->metrics.horiAdvance >> 6) * 1000 / pixel_size);
  if (full)
    {
      dwidth = glyph->bitmap.width;
      x = 0;
    }
  else
    {
      dwidth = glyph->metrics.horiAdvance >> 6;
      x = glyph->metrics.horiBearingX >> 6;
    }
  y = (glyph->metrics.horiBearingY - glyph->metrics.height) >> 6;
  printf ("DWIDTH %d 0\n", dwidth);
  printf ("BBX %d %d %d %d\n", glyph->bitmap.width, glyph->bitmap.rows, x, y);
  printf ("BITMAP\n");
  buf = (unsigned char *) glyph->bitmap.buffer;
  for (i = 0; i < glyph->bitmap.rows; i++)
    {
      for (j = 0; j < (glyph->bitmap.width + 7) / 8; j++)
	printf ("%02X", buf[i * glyph->bitmap.pitch + j]);
      printf ("\n");
    }
  printf ("ENDCHAR\n");
}


int
main (int argc, char **argv)
{
  FT_Library library;
  int err;
  int i;
  int pixel_size = DEFAULT_PIXEL_SIZE;
  FT_UInt code_table[0x10000];
  int nchars;
  char *filename;
  int platform_id, encoding_id;

  if (argc < 1)
    FATAL_ERROR ("Usage: %s ENCODING OTF-FILE\n", argv[0]);
  if (sscanf (argv[1], "%d-%d", &platform_id, &encoding_id) != 2)
    {
      platform_id = -1;
      filename = argv[1];
    }
  else
    filename = argv[2];
  
  if ((err = FT_Init_FreeType (&library)))
    FATAL_ERROR ("%s\n", "FT_Init_FreeType: error");
  err = FT_New_Face (library, filename, 0, &face);
  if (err == FT_Err_Unknown_File_Format)
    FATAL_ERROR ("%s\n", "FT_New_Face: unknown file format");
  else if (err)
    FATAL_ERROR ("%s\n", "FT_New_Face: unknown error");
  if (platform_id >= 0)
    {
      for (i = 0; i < face->num_charmaps; i++)
	if (face->charmaps[i]->platform_id == platform_id
	    && face->charmaps[i]->encoding_id == encoding_id)
	  break;
      if (i == face->num_charmaps)
	FATAL_ERROR ("Unknown ENCODING: %s\n", argv[1]);
      FT_Set_Charmap (face, face->charmaps[i]);
      if (platform_id == 0)
	{
	  if (encoding_id == 3)
	    registry = "iso10646-1";
	  else if (face->charmaps[i]->encoding_id == 4)
	    registry = "iso10646-full";
	}
      else if (face->charmaps[i]->platform_id == 3)
	{
	  if (face->charmaps[i]->encoding_id == 1)
	    registry = "iso10646-1";
	  else if (face->charmaps[i]->encoding_id == 10)
	    registry = "iso10646-full";
	}
      else
	{
	  registry = alloca (256);
	  sprintf (registry, "%d-%d", platform_id, encoding_id);
	}
    }
  else
    {
      registry = "raw-glyph";
    }

  {
    char *str = getenv ("PIXEL_SIZE");

    if (str && (i = atoi (str)) > 0)
      pixel_size = i;
  }

  if ((err = FT_Set_Pixel_Sizes (face, 0, pixel_size)))
    FATAL_ERROR ("%s\n", "FT_Set_Pixel_Sizes: error");

  /*
  for (i = nchars = 0; i < 0x10000; i++)
    if (! FT_Load_Glyph (face, i, FT_LOAD_RENDER | FT_LOAD_MONOCHROME)
	&& face->glyph->bitmap.rows * face->glyph->bitmap.width)
      nchars++;
  */
  for (i = 0; i < 0x10000; i++)
    {
      if (platform_id >= 0)
	{
	  code_table[i] = FT_Get_Char_Index (face, (FT_ULong) i);
	  if (! code_table[i])
	    {
	      code_table[i] = -1;
	      continue;
	    }
	}
      else
	{
	  code_table[i] = i;
	}
      if (! FT_Load_Glyph (face, code_table[i],
			   FT_LOAD_RENDER | FT_LOAD_MONOCHROME)
	  && face->glyph->bitmap.rows * face->glyph->bitmap.width)
	nchars++;
      else
	code_table[i] = -1;
    }

  dump_header (face, "unknown", nchars, pixel_size);
  for (i = 0; i < 0x10000; i++)
    if (code_table[i] >= 0)
      dump_image (pixel_size, code_table[i], i, 0);
  dump_tailer ();

  exit (0);
}


src/example/otflist.c
/* otflist.c -- List OpenType fonts.

Copyright (C) 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010
  National Institute of Advanced Industrial Science and Technology (AIST)
  Registration Number H15PRO167

This file is part of libotf.

Libotf is free software; you can redistribute it and/or modify it
under the terms of the GNU Lesser General Public License as published
by the Free Software Foundation; either version 2.1 of the License, or
(at your option) any later version.

Libotf is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library, in a file named COPYING; if not,
write to the Free Software Foundation, Inc., 59 Temple Place, Suite
330, Boston, MA 02111-1307, USA.  */

#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <sys/types.h>
#include <dirent.h>

#include <ft2build.h>
#include FT_FREETYPE_H

#include "config.h"
#ifdef HAVE_ALLOCA_H
#include <alloca.h>
#endif

#include <otf.h>

/* Format MSG by FMT and print the result to the stderr, and exit.  */

#define FATAL_ERROR(fmt, arg)	\
  do {				\
    fprintf (stderr, fmt, arg);	\
    exit (1);			\
  } while (0)


void
print_tag (OTF_Tag tag)
{
  char name[5];

  OTF_tag_name (tag, name);
  printf ("%s", name);
}

void
print_gsub_gpos_info (OTF *otf, char *table)
{
  int i, j;

  if (OTF_get_table (otf, table) == 0)
    {
      OTF_ScriptList *scripts;
      OTF_FeatureList *features;

      if (! strcmp (table, "GSUB"))
	scripts = &otf->gsub->ScriptList, features = &otf->gsub->FeatureList;
      else
	scripts = &otf->gpos->ScriptList, features = &otf->gpos->FeatureList;

      printf ("  %s scripts: ", table);
      for (i = 0; i < scripts->ScriptCount; i++)
	{
	  OTF_Script *script = scripts->Script + i;

	  if (i > 0)
	    printf (", ");
	  print_tag (script->ScriptTag);
	  if (script->LangSysCount > 0)
	    {
	      printf (" (");
	      for (j = 0; j < script->LangSysCount; j++)
		{
		  if (j > 0)
		    printf (", ");
		  print_tag (script->LangSysRecord[j].LangSysTag);
		}
	      printf (")");
	    }
	}
      printf ("\n");

      printf ("  %s features: ", table);
      for (i = 0; i < features->FeatureCount; i++)
	{
	  if (i > 0)
	    printf (",");
	  print_tag (features->Feature[i].FeatureTag);
	}
      printf ("\n");
    }
}
void
help_and_exit (char *prog)
{
  printf ("otflist %s\n", LIBOTF_VERSION);
  printf ("Usage: %s [-l] [-h] [DIR]\n", prog);
  printf ("List information about OpenType font files in the directory DIR.\n");
  printf ("It actually lists all fonts that can be handled by Freetype.\n");
  printf ("  -h		print this help, then exit\n");
  printf ("  -l		use a long listing mode\n");
  exit (0);
}

int
filter (const struct dirent *direntry)
{
  int len = strlen (direntry->d_name);
  const char *ext = direntry->d_name + (len - 4);

  return (len >= 5
	  && (! strncmp (ext, ".ttf", 4)
	      || ! strncmp (ext, ".TTF", 4)
	      || ! strncmp (ext, ".ttc", 4)
	      || ! strncmp (ext, ".TTC", 4)
	      || ! strncmp (ext, ".otf", 4)
	      || ! strncmp (ext, ".OTF", 4)
	      || ! strncmp (ext, ".PFA", 4)
	      || ! strncmp (ext, ".pfa", 4)
	      || ! strncmp (ext, ".PFB", 4)
	      || ! strncmp (ext, ".pfb", 4)));
}

int dir_index = 0;

#ifdef HAVE_SCANDIR

struct dirent **namelist = NULL;
int num_files = 0;

char *
next_file (char *dirname)
{
  if (dir_index == 0)
    {
#ifdef HAVE_ALPHASORT
      num_files = scandir (".", &namelist, filter, alphasort);
#else
      num_files = scandir (".", &namelist, filter, NULL);
#endif
    }
  if (dir_index == num_files)
    return NULL;
  return namelist[dir_index++]->d_name;
}

#else  /* not HAVE_SCANDIR */

DIR *dirp;

char *
next_file (char *dirname)
{
  struct dirent *dirent;

  if (dir_index == 0)
    dirp = opendir (dirname);
  while ((dirent = readdir (dirp))
	 && (strcmp (dirent->d_name, ".") == 0
	     || strcmp (dirent->d_name, "..") == 0));
  if (! dirent)
    return NULL;
  dir_index++;
  return dirent->d_name;
}

#endif /* not HAVE_SCANDIR */

int
main (int argc, char **argv)
{ 
  FT_Library ft_library;
  FT_Face face;
  char *filename;
  int long_format = 0;
  int i;

  if (FT_Init_FreeType (&ft_library))
    FATAL_ERROR ("%s\n", "!! Freetype initialization failed.");

  if (argc > 1)
    {
      if (! strcmp (argv[1], "-h") || ! strcmp (argv[1], "--help"))
	help_and_exit (argv[0]);
      if (! strcmp (argv[1], "-l"))
	long_format = 1, argc--, argv++;
    }
  if (argc == 2)
    {
      if (chdir (argv[1]) < 0)
	FATAL_ERROR ("Can't change directory to %s\n", argv[1]);
    }

  while ((filename = next_file (".")) != NULL)
    {
      if (! FT_New_Face (ft_library, filename, 0, &face))
	{
	  OTF *otf = OTF_open (filename);
	  char *name, *family = NULL, *style = NULL;

	  if (otf)
	    {
	      OTF_get_table (otf, "name");

	      if (! (family = otf->name->name[16]))
		family = otf->name->name[1];
	      if (! (style = otf->name->name[17]))
		style = otf->name->name[2];
	    }
	  if (! family)
	    family = face->family_name;
	  if (! style)
	    style = face->style_name;

	  name = alloca (strlen (filename)
			 + strlen (family)
			 + 4);
	  sprintf (name, "%s (%s)", filename, family);
	  printf ("%-40s %s", name, style);
	  for (i = 0; i < face->num_charmaps; i++)
	    printf (" %d-%d", face->charmaps[i]->platform_id,
		    face->charmaps[i]->encoding_id);
	  printf ("\n");
	  if (otf && long_format)
	    {
	      print_gsub_gpos_info (otf, "GSUB");
	      print_gsub_gpos_info (otf, "GPOS");
	    }
	  if (otf)
	    OTF_close (otf);
	}
      else
	{
	  printf ("%s fail to open\n", filename);
	}
    }
  exit (0);
}
